### PR TITLE
dynamic_broadcast_in_dim op spec

### PR DESCRIPTION
**Pull request by abhigunj**
Added Spec, Updated Verifier, ODS def and interpreter for the op.

Testing:
```
I1.  operand is type tensor or quantized tensor : ODS
I2. output_dimensions is  1-dimensional tensor of integer type : ODS 
I3. broadcast_dimensions is 1-dimensional constant tensor of integer type : ODS
I4. known_expanding_dimensions is 1-dimensional constant tensor of integer type : ODS
I5. known_non_expanding_dimensions is 1-dimensional constant tensor of integer type : ODS
C1.  element_type(result) = element_type(operand)  with special handling for per axis tensor : Verifier, test added
C2. size(broadcast_dimensions) = rank(operand) : Verifier, test added
C3.  0 <= broadcast_dimensions < rank(result) :  Verifier, test present
C4.  is_unique(broadcast_dimensions) : verifier, test added
C5.  ... : Verifier, Tests present
C6. ... :  Verifier, Test added
C7. .  size(output_dimensions) = rank(result) : Verifier, test present
C8, C9, C10. is_unique(known_expanding_dimensions + known_non_expanding_dimensions) :  Verifier, test added
C11, C12. 0 <= known_expanding_dimensions, known_non_expanding_dimensions  < rank(operand) : Verifier, test added  
```

Interpreter:
Supports static shapes only.

### CODE REVIEW

File: docs/spec.md [hunk 1]
```diff
@@ -2634,6 +2634,91 @@ planning to address this in
 
 &nbsp;[More Examples](https://github.com/openxla/stablehlo/tree/main/stablehlo/tests/interpret/dot_general.mlir)
 
+### dynamic_broadcast_in_dim
```
Comment (GleasonK) [inline docs/spec.md:2637]: Anything missing from `status.md`?

Comment (abhigunj) [reply docs/spec.md:2637]: thanks, updated op status.

File: docs/spec.md [hunk 2]
```diff
@@ -330,8 +330,8 @@ in StableHLO programs. In the meanwhile, here is the list of these operations:
   `torch_index_select`, `unary_einsum`
   ([#3](https://github.com/openxla/stablehlo/issues/3)).
 * "Dynamism" category of StableHLO operations - they were bootstrapped from
-   MHLO,and we are in the process of speccing them: `dynamic_broadcast_in_dim`,
-  `dynamic_conv`, `real_dynamic_slice`, `set_dimension_size`.
+   MHLO,and we are in the process of speccing them: `dynamic_conv`,
```
Comment (sdasgup3) [inline docs/spec.md:333]: we can remove dynamic_conv  now.

File: docs/spec.md [hunk 3]
```diff
@@ -2634,6 +2634,91 @@ planning to address this in
...
 
+### dynamic_broadcast_in_dim
+
+#### Semantics
+
+This operation is functionally identical to
+[broadcast_in_dim](https://github.com/openxla/stablehlo/blob/main/docs/spec.md#broadcast_in_dim)
+op, but the result shape is specified dynamically via `output_dimensions`.
+
+Op also accepts optional attributes `known_expanding_dimensions`, `known_non_expanding_dimensions`
```
Comment (sdasgup3) [inline docs/spec.md:2645]: The operation also...

File: docs/spec.md [hunk 4]
```diff
@@ -2634,6 +2634,91 @@ planning to address this in
...
+    zero_points(operand)[0] for i in
+    range(dim(result, quantization_dimension(result)))`.
+* (C7) `size(output_dimensions) = rank(result)`.
+* (C8) `is_unique(known_expanding_dimensions + known_non_expanding_dimensions)`.
+* (C9) `0 <= known_expanding_dimensions, known_non_expanding_dimensions  < rank(operand)`.
+
+#### Examples
+
+```mlir
+// %operand: [
```
Comment (sdasgup3) [inline docs/spec.md:2692]:
```
// %operand: [
//            [1, 2, 3]
//           ]
```

same as broadcast_in_dim

File: stablehlo/dialect/StablehloOps.td [hunk 5]
```diff
@@ -1951,25 +1951,36 @@ def StableHLO_DynamicBroadcastInDimOp : StableHLO_ShapedInterfaceOp<
...
     must be disjoint and they must be a subset of the operand's dimensions.
+
+    See: https://github.com/openxla/stablehlo/blob/main/docs/spec.md#dynamic_broadcast_in_dim
+
+    Example:
+    ```mlir
+    %output_dimensions = stablehlo.constant dense<[3, 2, 2]> : tensor<3xi64>
+    %result = "stablehlo.dynamic_broadcast_in_dim"(%operand, %output_dimensions) {
+      broadcast_dimensions = array<i64: 0, 2>,
+      known_expanding_dimensions = array<i64: 0>,
```
Comment (sdasgup3) [inline stablehlo/dialect/StablehloOps.td:1971]: 0 is non exampling, 1 is expanding

Comment (sdasgup3) [reply stablehlo/dialect/StablehloOps.td:1971]:
```
  known_expanding_dimensions = array<i64: 1>,
  known_non_expanding_dimensions = array<i64: 0>
```

File: stablehlo/tests/interpret/dynamic_broadcast_in_dim.mlir [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+// %operand: [
+//            [1, 2, 3]
+//           ]
+func.func @broadcast_in_dim() {
+  %operand = stablehlo.constant dense<[[1, 2, 3]]> : tensor<1x3xi64>
+  %output_dimensions = stablehlo.constant dense<[2, 3, 2]> : tensor<3xi64>
+  %result = "stablehlo.dynamic_broadcast_in_dim"(%operand, %output_dimensions) {
+    broadcast_dimensions = array<i64: 2, 1>,
+    known_expanding_dimensions = array<i64: 0>,
+    known_non_expanding_dimensions = array<i64: 1>
```
Comment (sdasgup3) [inline stablehlo/tests/interpret/dynamic_broadcast_in_dim.mlir:12]:
```  
known_expanding_dimensions = array<i64: 1>,
  known_non_expanding_dimensions = array<i64: 0>
```

File: stablehlo/tests/interpret/dynamic_broadcast_in_dim.mlir [hunk 7]
```diff
@@ -0,0 +1,28 @@
+// RUN: stablehlo-translate --interpret -split-input-file %s
+
+// %operand: [
```
Comment (sdasgup3) [inline stablehlo/tests/interpret/dynamic_broadcast_in_dim.mlir:3]: remove the comment?

File: stablehlo/dialect/TypeInference.cpp [hunk 8]
```diff
@@ -310,6 +310,41 @@ LogicalResult verifyReshapeOpQuantizationConstraints(
...
+    ArrayRef<int64_t> broadcastDimensions) {
+  auto operandType = cast<RankedTensorType>(operand.getType());
+  auto resultType = cast<RankedTensorType>(result.getType());
+  auto resultQType = cast<quant::UniformQuantizedPerAxisType>(
+      getElementTypeOrSelf(result.getType()));
+  auto operandQType = cast<quant::UniformQuantizedPerAxisType>(
+      getElementTypeOrSelf(operand.getType()));
+  auto operandQDim = operandQType.getQuantizedDimension();
+  auto resultQDim = resultQType.getQuantizedDimension();
+  // broadcast_in_dimm_c6, dynamic_broadcast_in_dimm_c6
```
Comment (mlevesquedion) [inline stablehlo/dialect/TypeInference.cpp:324]: dimm -> dim

File: stablehlo/dialect/TypeInference.cpp [hunk 9]
```diff
@@ -310,6 +310,41 @@ LogicalResult verifyReshapeOpQuantizationConstraints(
...
+  if (resultQDim != broadcastDimensions[operandQDim])
+    return emitOptionalError(location, "result quantization_dimension ",
+                             resultQDim, " not same as broadcast_dimensions[",
+                             operandQDim,
+                             "] = ", broadcastDimensions[operandQDim]);
+  if (operandType.getDimSize(operandQDim) == 1) {
+    for (int64_t j = 0; j != resultType.getDimSize(resultQDim); ++j) {
```
Comment (mlevesquedion) [inline stablehlo/dialect/TypeInference.cpp:331]: Nit: use `i` instead of `j` (convention)

**Review by mlevesquedion — APPROVED** [review]
LGTM, just a few nits.

File: docs/spec.md (hunk 2 above)
Comment (abhigunj) [reply docs/spec.md]: done

**Review by ghpvnist — CHANGES_REQUESTED** [review]

File: docs/spec.md (hunk 4 above)
Comment (abhigunj) [reply docs/spec.md:2692]: I used other example vs broadcast so for this is correct for 1x3 tensor. 

after reconsidering,  keeping the spec example same as broadcast_in_dim for consistency.  Taken care in the latest draft.

File: stablehlo/tests/interpret/dynamic_broadcast_in_dim.mlir (hunk 6 above)
Comment (abhigunj) [reply stablehlo/tests/interpret/dynamic_broadcast_in_dim.mlir]:
```
 known_expanding_dimensions = array<i64: 0>,
  known_non_expanding_dimensions = array<i64: 1>

```
is correct for 
```
// %operand: [
//            [1, 2, 3]
//           ] 

to 

// %result: [
//            [
//             [1, 1],
//             [2, 2],
//             [3, 3]
//            ],
//            [
//             [1, 1],
//             [2, 2],
//             [3, 3]
//            ]
//          ]
```
operand dim 0 size = 1
result dim 0 size = 2

File: stablehlo/dialect/StablehloOps.td (hunk 5 above)
Comment (abhigunj) [reply stablehlo/dialect/StablehloOps.td:1977]: corrected. thank you.

File: stablehlo/dialect/TypeInference.cpp (hunk 8 above)
Comment (abhigunj) [reply stablehlo/dialect/TypeInference.cpp]: it came from https://en.wikipedia.org/wiki/DIMM ;) 

corrected, thanks.

File: stablehlo/tests/ops_stablehlo.mlir [hunk 10]
```diff
@@ -1017,73 +1018,71 @@ func.func @dynamic_broadcast_in_dim_ok_dim(%arg0: tensor<1xf32>, %shape: tensor<
...
+func.func @dynamic_broadcast_in_dim_c2(%arg0: tensor<?x?xi32>, %shape: tensor<3xi64>) -> tensor<?x?x?xi32> {
+  // expected-error@+1 {{broadcast_dimensions size (1) does not match operand rank (2)}}
+  %0 = "stablehlo.dynamic_broadcast_in_dim"(%arg0, %shape) {broadcast_dimensions = array<i64: 1>} : (tensor<?x?xi32>, tensor<3xi64>) -> tensor<?x?x?xi32>
+  func.return %0 : tensor<?x?x?xi32>
 }
 
 // -----
 
-func.func @dynamic_broadcast_in_dim_negative_size(%arg0: tensor<1xf32>, %shape: tensor<3xi64>) -> tensor<7x8x9xf32> {
+func.func @dynamic_broadcast_in_dim_c3_negative_size(%arg0: tensor<1xf32>, %shape: tensor<3xi64>) -> tensor<7x8x9xf32> {
```
[the remark this answers is not in the document]
Comment (abhigunj) [reply stablehlo/tests/ops_stablehlo.mlir:1053]: done

File: stablehlo/dialect/TypeInference.cpp [hunk 11]
```diff
@@ -3826,16 +3849,19 @@ LogicalResult verifyDynamicBroadcastInDimOp(
       };
   collectExpansionBehaviorDims(knownExpandingDimensions);
   collectExpansionBehaviorDims(knownNonexpandingDimensions);
+
+  // dynamic_broadcast_in_dim_c8
   if (knownExpansionBehavior.size() != numKnownExpansionBehavior)
     return emitOptionalError(
         location,
         "duplicate expansion hint for at least one operand dimension");
```
[the remark this answers is not in the document]
Comment (abhigunj) [reply stablehlo/dialect/TypeInference.cpp:3998]: this check is common for both the attributes, keeping the log message unchanged as discuss.  Added a marker here.

**Review by ghpvnist — APPROVED** [review]

**Pull request merged**